### PR TITLE
Adding version 0.4.6 of simdjson (up from just having version 0.2.1)

### PR DESCRIPTION
**Pull request by lemire**
Specify library name and version:  **simdjson/0.4.6**

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/simdjson/0.4.6/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+               "fPIC": [True, False],
+               "threads": [True, False]}
+    default_options = {'shared': False,
+                       'fPIC': True,
+                       'threads': True}
+    _source_subfolder = "source_subfolder"
+
+    @property
+    def _supported_cppstd(self):
+        return ["11", "gnu11", "17", "gnu17", "20", "gnu20"]
```
Comment (uilianries) [inline recipes/simdjson/0.4.6/conanfile.py:28]: What about 14?

File: recipes/simdjson/0.4.6/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    def _has_support_for_cpp11(self):
+        supported_compilers = [("apple-clang", 10), ("clang", 6), ("gcc", 7), ("Visual Studio", 15.7)]
+        compiler, version = self.settings.compiler, Version(self.settings.compiler.version)
+        return any(compiler == sc[0] and version >= sc[1] for sc in supported_compilers)
+
+    def configure(self):
+        if self.settings.compiler == "Visual Studio":
+            self.options.remove("fPIC")
```
Comment (uilianries) [inline recipes/simdjson/0.4.6/conanfile.py:37]:
```suggestion
            del self.options.fPIC
```

File: recipes/simdjson/all/CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +0,0 @@
-cmake_minimum_required(VERSION 3.9)
```
Comment (uilianries) [inline recipes/simdjson/all/CMakeLists.txt:1]: Need to be replicated to 0.4.6. conan_basic_setup is required

File: recipes/simdjson/0.4.6/conanfile.py (hunk 1 above)
Comment (lemire) [reply recipes/simdjson/0.4.6/conanfile.py]: We just support C++11 and C++17. None of that C++14 crap. (Joke.)

Fixed.

File: recipes/simdjson/0.4.6/conanfile.py (hunk 2 above)
Comment (lemire) [reply recipes/simdjson/0.4.6/conanfile.py]: Done.

File: recipes/simdjson/all/CMakeLists.txt (hunk 3 above)
Comment (lemire) [reply recipes/simdjson/all/CMakeLists.txt:1]: Can you elaborate? There is a file named recipes/simdjson/0.4.6/test_package/CMakeLists.txt that includes conan_basic_setup.

Comment (lemire) [reply recipes/simdjson/all/CMakeLists.txt]: Oh. I figured it out.

Comment (lemire) [reply recipes/simdjson/all/CMakeLists.txt]: Fixed.

**Comment by lemire** [comment]
@uilianries Thanks for the review. Can you look again and tell me if it is better?

File: recipes/simdjson/0.4.6/conanfile.py (hunk 1 above)
Comment (uilianries) [reply recipes/simdjson/0.4.6/conanfile.py]: Only because it's a minor :joy:

File: recipes/simdjson/0.4.6/conanfile.py (hunk 2 above)
Comment (uilianries) [reply recipes/simdjson/0.4.6/conanfile.py]: I didn't explain, but `self.options.remove("fPIC")` has been deprecated. Once approved in EAP, the CCI bot will check your recipe and decline.

del is more pythonic, just like delete in C++, so we don't re-implement the wheel

File: recipes/simdjson/0.4.6/conanfile.py (hunk 1 above)
Comment (lemire) [reply recipes/simdjson/0.4.6/conanfile.py]: Anyhow, fixed in a follow-up commit.

File: recipes/simdjson/0.4.6/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.9)
+project(test_package)
+
+set(CMAKE_VERBOSE_MAKEFILE TRUE)
```
Comment (uilianries) [inline recipes/simdjson/0.4.6/test_package/CMakeLists.txt:4]:
```suggestion
```

CCI bot will block it.

I used to enable it when building on my machine, but for regular builds, it only increases the build log and people don't read.

File: recipes/simdjson/0.4.6/conanfile.py (hunk 2 above)
Comment (lemire) [reply recipes/simdjson/0.4.6/conanfile.py]: That's fine. I am fluent in Python and I know about del.

Please keep the comments coming!

File: recipes/simdjson/0.4.6/test_package/CMakeLists.txt (hunk 4 above)
Comment (lemire) [reply recipes/simdjson/0.4.6/test_package/CMakeLists.txt]: Done. Fixed in a follow-up commit.

**Comment by lemire** [comment]
> I would like to unify both versions at same folder all, but I see some things really changed, including the test_package, thus keeping both versions separated looks right.

The API changed. The old API is still available, but the new one is better.

Also the requirements have changed. The new simdjson is much more generic.

**Comment by uilianries** [comment]
@lemire it looks great now. Let's wait for EAP approval. It happens in waves, you will be notified when it happens there in #4.

**Comment by lemire** [comment]
@uilianries Can you help? I don't understand the error message.

**Comment by danimtb** [comment]
@lemire as this library has two different recipes depending on the version, you would have to split the changes in two different PRs as the build service can only build one reference at a time. Thank you!

**Comment by lemire** [comment]
@danimtb 

Moved part of the PR to https://github.com/conan-io/conan-center-index/pull/2237

File: recipes/simdjson/config.yml [hunk 5]
```diff
@@ -1,3 +1,5 @@
 versions:
   "0.2.1":
-    folder: all
+    folder: 0.2.1
+  "0.4.6":
+    folder: 0.4.6
```
Comment (prince-chrismc) [inline recipes/simdjson/config.yml:5]:
```suggestion
    folder: 0.4.6

```

File: recipes/simdjson/0.4.6/conandata.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+sources:
+  "0.4.6":
+    url: "https://github.com/lemire/simdjson/archive/v0.4.6.tar.gz"
+    sha256: "2e6e708a5973eef7442ab19b3fe76bdd791cdc9b67020ca1190e00832517808b"
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/conandata.yml:4]:
```suggestion
    sha256: "2e6e708a5973eef7442ab19b3fe76bdd791cdc9b67020ca1190e00832517808b"

```

File: recipes/simdjson/0.4.6/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.9)
+
+project(cmake_wrapper)
+
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
+conan_basic_setup()
+
+add_subdirectory(source_subfolder)
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/CMakeLists.txt:8]:
```suggestion
add_subdirectory(source_subfolder)

```

File: recipes/simdjson/0.4.6/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        # remove unneeded directories
+        tools.rmdir(os.path.join(self.package_folder, 'lib', 'cmake'))
+        tools.rmdir(os.path.join(self.package_folder, 'lib', 'pkgconfig'))
+
+        self.copy("license", src=self._source_subfolder, dst="licenses", ignore_case=True, keep_path=False)
+
+    def package_info(self):
+        self.cpp_info.libs = ['simdjson']
+        if self.settings.os == "Linux":
+            self.cpp_info.system_libs = ["m"]
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/conanfile.py:79]: Possibly? 

```suggestion
            self.cpp_info.system_libs = ["m"]
            if self.options.threads:
                self.cpp_info.system_libs = ["pthreads"]
```

File: recipes/simdjson/0.4.6/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    exports_sources = ["CMakeLists.txt"]
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False],
+               "fPIC": [True, False],
+               "threads": [True, False]}
+    default_options = {'shared': False,
+                       'fPIC': True,
+                       'threads': True}
+    _source_subfolder = "source_subfolder"
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/conanfile.py:24]: namely to cache CMake (configuring twice is expensive)

```suggestion
     _cmake = None
     
    @property
    def _source_subfolder(self):
          return "source_subfolder"
```

File: recipes/simdjson/0.4.6/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions['SIMDJSON_BUILD_STATIC'] = not self.options.shared
+        cmake.definitions['SIMDJSON_ENABLE_THREADS'] = self.options.threads
+        cmake.definitions['SIMDJSON_SANITIZE'] = False
+        cmake.definitions['ENABLE_FUZZING'] = False
+        cmake.configure(source_folder=self._source_subfolder)
+        return cmake
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/conanfile.py:61]: Part 2 of caching cmake

```suggestion
        if(self._cmake):
                return self._cmake
        self._cmake = CMake(self)
        self._cmake.definitions['SIMDJSON_BUILD_STATIC'] = not self.options.shared
        self._cmake.definitions['SIMDJSON_ENABLE_THREADS'] = self.options.threads
        self._cmake.definitions['SIMDJSON_SANITIZE'] = False
        self._cmake.definitions['ENABLE_FUZZING'] = False
        self._cmake.configure(source_folder=self._source_subfolder)
        return self._cmake
```

File: recipes/simdjson/0.4.6/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                       'fPIC': True,
+                       'threads': True}
+    _source_subfolder = "source_subfolder"
+
+    @property
+    def _supported_cppstd(self):
+        return ["11", "gnu11","14", "gnu14", "17", "gnu17", "20", "gnu20"]
+
+    def _has_support_for_cpp11(self):
+        supported_compilers = [("apple-clang", 10), ("clang", 6), ("gcc", 7), ("Visual Studio", 15.7)]
```
Comment (prince-chrismc) [inline recipes/simdjson/0.4.6/conanfile.py:31]: c++11 is supported on lower compiler versions, the combination seems odd :thinking: Would in be possible to get some insight I'm curious?

Si je retroune au universite, ce sera pour une de tes cours!

**Comment by prince-chrismc** [comment]
This was 'merged' with out all the checks passing. does not look like the new version is in master though :confused: 

@SSE4 you may have confused GitHub

**Comment by lemire** [comment]
@prince-chrismc  I should have closed this PR when I opened the new one. I will open a new PR with 0.4.6 after taking into account the necessary changes. It will include this PR as a subset so it should fix the issues.

**Comment by lemire** [comment]
@prince-chrismc I have created a new PR.

**Pull request merged**